include/xtensor-io/xio_zlib.hpp

```diff
@@ -0,0 +1,347 @@
+/***************************************************************************
+* Copyright (c) Wolf Vollprecht, Sylvain Corlay and Johan Mabille          *
+* Copyright (c) QuantStack                                                 *
+*                                                                          *
+* Distributed under the terms of the BSD 3-Clause License.                 *
+*                                                                          *
+* The full license is in the file LICENSE, distributed with this software. *
+****************************************************************************/
+
+#ifndef XTENSOR_IO_ZLIB_HPP
+#define XTENSOR_IO_ZLIB_HPP
+
+#include <fstream>
+
+#include "zlib.h"
+
+#include "xtensor/xadapt.hpp"
+#include "xtensor-io.hpp"
+#include "xfile_array.hpp"
+#include "xio_stream_wrapper.hpp"
+
+#define ZLIB_CHUNK 0x4000
+
+namespace xt
+{
+    namespace detail
+    {
+        inline std::string zlib_err(int ret)
+        {
+            switch (ret)
+            {
+                case Z_STREAM_ERROR:
+                    return "invalid compression level";
+                case Z_DATA_ERROR:
+                    return "invalid or incomplete deflate data";
+                case Z_MEM_ERROR:
+                    return "out of memory";
+                case Z_VERSION_ERROR:
+                    return "zlib version mismatch";
+                default:
+                    return "unknown";
+            }
+        }
+
+        template <typename T, class I>
+        inline xt::svector<T> load_zlib(I& stream, bool as_big_endian)
+        {
+            xt::svector<T> uncompressed_buffer;
+            int ret;
+            unsigned have;
+            z_stream strm;
+            unsigned char in[ZLIB_CHUNK];
+            T out[ZLIB_CHUNK / sizeof(T)];
+
+            strm.zalloc = Z_NULL;
+            strm.zfree = Z_NULL;
+            strm.opaque = Z_NULL;
+            strm.avail_in = 0;
+            strm.next_in = Z_NULL;
+            ret = inflateInit(&strm);
+            if (ret != Z_OK)
+            {
+                XTENSOR_THROW(std::runtime_error, "zlib decompression failed (" + zlib_err(ret) + ")");
+            }
+            do
+            {
+                stream.read((char*)in, sizeof(in));
```

Comment:
I think the array `in` should automatically decay to `char*` here, so no need for an explicit cast.

Reply:
If I remove the cast I get `error: invalid conversion from ‘unsigned char*’ to ‘char*’`.

Reply:
Then in should probably an array of `char` instead of `unsigned char`.

Reply:
:+1: